### ISSUE TITLE
JS driver doesn't reject invalid nonce from the server

### ISSUE THREAD

**Issue by neumino** [rethinkdb/rethinkdb#5916]
This [line](https://github.com/rethinkdb/rethinkdb/blob/next/drivers/javascript/net.coffee#L1145) 
```
if not (auth_r.substr(0, r_string.length) == r_string)
```
is in the npm package:
```
if (!auth_r.substr(0, r_string) === r_string) {
```
When it should be (note the extra parentheses):
```
if (!(auth_r.substr(0, r_string) === r_string)) {
```

The coffeescript playground is generating the appropriate code, so I guess the coffee script version used to generate the npm package has a bug?

**Comment by neumino** [comment]
Err, one more thing. It should be `auth_r.substr(0, r_string.length) ` and not `auth_r.substr(0, r_string)`

**Comment by neumino** [comment]
Also, [this line](https://github.com/rethinkdb/rethinkdb/blob/next/drivers/javascript/net.coffee#L1079) should be `result |= a[i] ^ b[i]`. `xor_bytes` returns a buffer, not a number.

**Comment by danielmewes** [comment]
Thanks for the report @neumino ! Good catch.

@nighelles Could you look into the missing `.length`?
I'll try to figure out which coffeescript version we need to build with to avoid the wrong translation.

**Comment by danielmewes** [comment]
Actually it turns out this is a bug in our code, which was fixed in `next`, but not in `v2.3.x`.
The line in next:
https://github.com/rethinkdb/rethinkdb/blob/next/drivers/javascript/net.coffee#L1145
The line in v2.3.x:
https://github.com/rethinkdb/rethinkdb/blob/v2.3.x/drivers/javascript/net.coffee#L1142

It was fixed by https://github.com/rethinkdb/rethinkdb/commit/5bb33f80f729053874be8f2df83983a1e098c7d2 but apparently not backported. I'll double-check that nothing else is missing.

**Comment by danielmewes** [comment]
Cherry-picked via 6d32cc1. Another missing commit was https://github.com/rethinkdb/rethinkdb/commit/651588fafe1fc19d066bc6714c3f09e4c119451b , which is now cherry-picked into v2.3.x as well via 94c0465.

**Comment by danielmewes** [comment]
Re-opening because the `xor_bytes` thing is still fishy (thought it does appear to work somehow).

**Comment by danielmewes** [comment]
A fix for the `xor_bytes` misuse issue is now in code review 3718.

**Comment by danielmewes** [comment]
The `compare_digest` / `xor_bytes` issue is fixed in v2.3.x d167790 and next 6d19d93.
